autometa/common/utilities.py

```diff
@@ -122,17 +128,20 @@ def gunzip(infpath, outfpath):
         `outfpath` already exists and is not empty
 
     """
-    logger.debug(
-        f'gunzipping {os.path.basename(infpath)} to {os.path.basename(outfpath)}')
-    if os.path.exists(outfpath) and os.stat(outfpath).st_size > 0:
+    logger.debug(f'gunzipping {os.path.basename(infpath)} to {os.path.basename(outfpath)}')
+    if os.path.exists(outfpath) and os.path.getsize(outfpath) > 0:
         raise FileExistsError(outfpath)
     lines = ''
-    with gzip.open(infpath) as fh:
-        for line in fh:
-            lines += line.decode()
-    with open(outfpath, 'w') as out:
-        out.write(lines)
+    with gzip.open(infpath, 'rt') as fh, open(outfpath, 'w') as out:
+        for i,line in enumerate(fh):
+            lines += line
+            if i >= block_size:
+                out.write(lines)
+                lines = ''
     logger.debug(f'gunzipped {infpath} to {outfpath}')
+    if delete_original:
+        os.remove(infpath)
+        logger.debug(f'removed original file: {infpath}')
     return outfpath
```

Comment:
This function should call `write_checksum` so that the file it produces can be checked later.

Reply:
Do you mean only when the function is instructed to remove the original file?

Reply:
No. After it makes the `out` file (successfully), it should call `write_checksum`. Probably should verify that there is a checksum file there for the original archive as well before using.

Reply:
1. I'm not sure this is appropriate in the context of `gunzip`. It seems like this behavior should only be performed if specified.

2. Are we planning on hosting checksums for the database files that _do not_ have hosted checksums?

3. In relation to issue-#71, I was thinking of writing a general function as a decorator, so when we reach a step in the pipeline we would like Autometa to check, we can wrap the step with `@checkpoint` or some equivalent. Perhaps this would take as an argument (as well as the function) an output file path, such as `@utilities.checkpoint(out=outfpath)`. We would similarly require the function have an output file path argument to avoid confusion.

Example:

```python
@checkpoint
def some_func(infpath, out):
    """"This function takes as input an `infpath` and writes an `out`""""
    some_info = read(infpath)
    write(outfpath)
    return outfpath

# Since `@checkpoint` is specified
# `out` will be parsed from the function and checksummed
# then added to the checkpoints file for lookup later
```

I need to work on this implementation still, but for the future, just being able to place a `@checkpoint` decorator on a step in the pipeline that writes a file seems much simpler than writing out `checkpoint` functionality on a function-by-function basis.

Reply:
> No. After it makes the `out` file (successfully), it should call `write_checksum`. Probably should verify that there is a checksum file there for the original archive as well before using.

Do you mean given `infpath`, look for `f'{infpath}.md5'`? If this does not exist, does this really tell us anything?

Reply:
Well, if you commit to always write a md5 file after any output step finishes without error, then it is an easy check later that you did that step, and you can then go ahead and check the file against the checksum. The two checks are:

1. md5 file exists - therefore step completed successfully at some point in the past
2. md5 checksum in the md5 file matches that which I just calculated with the file - therefore the file has not been messed with or corrupted since it was made.

Reply:
> Well, if you commit to always write a md5 file after any output step finishes without error, then it is an easy check later that you did that step, and you can then go ahead and check the file against the checksum. The two checks are:
> 
> 1. md5 file exists - therefore step completed successfully at some point in the past

We also discussed writing the checksums to one file at the completion of any output step here https://github.com/KwanLab/Autometa/issues/71#issuecomment-629479469. Do you think we should do both? There would be some redundant behavior. One file may reduce the clutter in the output directory as well as eliminate the need to perform multiple `file exists` checks.  Basically the commitment would be to write the checksum to the checkpoints file at each successful step rather than a written checksum at each step.

The analogous check from point 1. would be checking that the named output is in the checkpoints file. If it is, then we can retrieve the checksum value for comparison against the calculated checksum of the file (point 2).

Reply:
I don't think it matters conceptually, but there is a sort of "all eggs in one basket" problem with using one file - if that file gets corrupted then you are screwed. I mean, it won't be a big file so perhaps this is taking paranoia to a whole new level. Just thought I'd put that out there.

Reply:
Okay. I suppose it does not hurt to have an extra fail safe. Good point with the "all eggs in one basket", this makes sense. Currently the database files are being written with checksums. It sounds like for this, we pretty much have all of the functionality we need just from `write_checksum`, `calc_checksum` and `read_checksum`.